### PR TITLE
EulerDiscreteScheduler add `rescale_betas_zero_snr`

### PR DESCRIPTION
**Pull request by Beinsezii**
### Overview
Adds support for the `rescale_betas_zero_snr` config option to EulerDiscreteScheduler. Currently it works equivelently to DDIM with the concession that the final alpha_cumprod is patched to 2 ** -24 to resolve the `inf` issue, similar to what ComfyUI does.

~~This does not follow the 'curve' so to speak, and a value closer to 0 would be more appropriate, however lower values such as 2 ** -24 run into precision issues on fp16/bf16 inference. A more proper fix would be finding exactly where the precision issues lie in the pipeline code and resolving them (upcast?).  don't think I am math enough for this; 2 ** -16 is satisfactory from my benchmark images on the model "ptx0/terminus-xl-gamma-training".~~

I have fixed the precision issues by upcasting the sample for the duration of the step() function, then re-casting back to the model's dtype before return. There appears to be no performance loss.

### Examples
Unpatched EulerDiscrete
unpatched

DDIM as reference
ddim

Patched EulerDiscrete

### Uncertainties
 - The function `rescale_zero_terminal_snr` could reasonably be moved to `scheduling_utils.py` and shared between EulerDiscrete and DDIM, however with convention shown via `betas_for_alpha_bar` I've opted to duplicate it. If I or someone eventually adds ZSNR support to other samplers, this may be desirable.

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 1]
```diff
@@ -92,6 +92,42 @@ def alpha_bar_fn(t):
     return torch.tensor(betas, dtype=torch.float32)
 
 
+def rescale_zero_terminal_snr(betas):
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_euler_discrete.py:95]: is this copied from ddim? if so let's add a # Copy from statement

**Review by yiyixuxu — APPROVED** [review]
thanks! left a comment about `#Copies from` statement
and +1 on adding a test

File: src/diffusers/schedulers/scheduling_euler_discrete.py (hunk 1 above)
Comment (Beinsezii) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:95]: It is copied from DDIM which itself is copied 1:1 (and renamed) from Page 3 Alrogithm 1 of [the ZSNR paper](https://arxiv.org/pdf/2305.08891.pdf).

**Comment by Beinsezii** [comment]
Added the `Copy` comment and rescale test. `pytest` passed it locally so I think that'll be it once the CI checks finish.

**Comment by Beinsezii** [comment]
Yea you can see it in my ddim reference image too. Compared to the patches euler, it has a lot of residual noise. That's one of my main motivations for porting the beta rescale.

**Comment by patrickvonplaten** [comment]
@yiyixuxu feel free to merge whenever

**Review by yiyixuxu — APPROVED** [review]
thanks!
